active/0000-anonymous-amd.md

```diff
@@ -0,0 +1,98 @@
+- Start Date: (fill me in with today's date, 2016-06-14)
+- RFC PR: (leave this empty)
+- ember-cli Issue: (leave this empty)
+
+# Summary
+
+This RFC proposes extending the `app.import` API to consume anonymous AMD modules. It accompanies [ember-cli PR 5976](https://github.com/ember-cli/ember-cli/pull/5976).
+
+# Motivation
+
+AMD modules come in two flavors: named and anonymous. _Anonymous AMD_ is the more portable distribution format, but it typically requires preprocessing into _named AMD_ before it can be included into an application.
+
+    /* Anonymous AMD Examples */
+
+    // direct value
+    define({ color: 'black' });
+
+    // function returning value
+    define(function() { return { color: 'black' }; });
+
+    // function returning value, with declared dependencies
+    define(["jquery", "moment"], function(jQuery, moment) {
+      return {
+        injectTime: function() {
+          jQuery('#time-box').html(moment().format('HH:MM'));
+        }
+      }
+    });
+
+    /* Named AMD Examples */
+
+    // direct value
+    define('my-config', { color: 'black' });
+
+    // function returning value
+    define('my-config', function() { return { color: 'black' }; });
+
+    // function returning value, with declared dependencies
+    define('time-utils', ["jquery", "moment"], function(jQuery, moment) {
+      return {
+        injectTime: function() {
+          jQuery('#time-box').html(moment().format('HH:MM'));
+        }
+      }
+    });
+
+
+Today, ember-cli users can add arbitrary third-party _named AMD_ modules into their application via:
+
+    app.import('/path/to/module.js');
+
+But this does not support _anonymous AMD_ modules, which is annoying because _anonymous AMD_ is the better format for library distribution and is widely used.
+
+# Detailed design
+
+In order to de-anonymize AMD, it's necessary to choose a name for the module for use within a given application. So I propose extending the:
+
+    app.import('/path/to/module.js');
+
+API with an additional argument:
+
+    app.import('/path/to/module.js', {
+      using: [
+        { transformation: 'amd', as: 'some-dep' }
+      ]
+    });
+
+`using` provides a list of transformations. Each transformation is identified by its `transformation` property. Any other properties are treated as arguments to the transformation implementation -- they are opaque to ember-cli. Transformations will run in the given order.
```

Comment:
We didn't actually implement this API in our current implementation. The loop that we run here iterates over the transforms, errors if it isn't AMD, and then sets the last one:
https://github.com/ember-cli/ember-cli/blob/e9a29f8246340a60b585f75825b2a64ae9001b72/lib/broccoli/ember-app.js#L1563-L1577

```javascript
app.import('vendor/anonymous-amd-example.js', {
  using: [
    { transformation: 'amd',  as: 'hello-world'},
    { transformation: 'amd',  as: 'this-is-dumb-but-demonstrates-the-bug'}
  ],
  outputFile: '/assets/output.js'
});
```

We then process the resulting pre-clobbered `amdModuleNames` property without regards for order of any additional transformations, or even aware of the transformations array at all:
https://github.com/ember-cli/ember-cli/blob/e9a29f8246340a60b585f75825b2a64ae9001b72/lib/broccoli/ember-app.js#L1005-L1014

/cc @ef4 @stefanpenner